Compiler Stack Depth Error with Foundry Nightly/Stable, Works on v0.3.0

I'm encountering a compilation error when using the **nightly** or **`stable`** version of Foundry via `foundryup --tag stable` or `--tag nightly`, while everything works fine with **v0.3.0**, which is also what the GitHub build bot uses.

#### ✅ Works:

```bash
foundryup --tag v0.3.0
forge build  # ✅ successful
```

#### ❌ Fails on stable/nightly:

```bash
foundryup --tag stable
forge build
```

Error:

```
[⠔] Solc 0.8.25 finished in 1.40s
Error: Compiler run failed:
Error: Yul exception:Cannot swap Slot RET[cleanup_uint256] with Variable expr_48931_component_5_mpos: too deep in the stack by 4 slots in [ var_rateQueue_slot RET[cleanup_uint256] var_rail_48809_slot var_skipArbitration var_targetEpoch RET var_railId var_segmentEndBoundary RET[cleanup_uint256] 0x08 var_currentRate var_totalOperatorCommission var_totalPaymentFee var_rail_48809_slot var_totalSettledAmount RET[read_from_storage_split_offset_uint256] var_processedEpoch expr_48931_component expr_48931_component_1 expr_48931_component_2 expr_48931_component_3 expr_48931_component_5_mpos ]
memoryguard was present.
make: *** [build] Error 1
```

---

### 💭 Thoughts

It seems like something in the latest Foundry tooling or the Yul optimizer path introduced a breaking change — possibly increased sensitivity to stack depth or a regression in the IR pipeline.

While it's acceptable to pin to `v0.3.0`, this may create issues for developers or downstream projects expecting compatibility with the latest `foundryup` versions.

---

### ✅ Potential Fixes

1. **Update the Makefile** to explicitly install `v0.3.0` via `foundryup --tag v0.3.0`
2. **Document the Foundry version requirement** (v0.3.0) in `README.md` or build instructions
3. **Investigate & resolve stack-depth issue** in newer Foundry builds / IR passes

--- a/src/Payments.sol
+++ b/src/Payments.sol
@@ -139,6 +139,15 @@ contract Payments is
     // Tracks whether a token has ever had fees collected, to prevent duplicates in feeTokens
     mapping(address => bool) public hasCollectedFees;
 
+    struct SettlementState {
+        uint256 totalSettledAmount;
+        uint256 totalNetPayeeAmount;
+        uint256 totalPaymentFee;
+        uint256 totalOperatorCommission;
+        uint256 processedEpoch;
+        string note;
+    }
+
     /// @custom:oz-upgrades-unsafe-allow constructor
     constructor() {
         _disableInitializers();
@@ -955,7 +964,7 @@ contract Payments is
             uint256 totalOperatorCommission,
             uint256 finalSettledEpoch,
             string memory note
-        )
+        )   
     {
         return settleRailInternal(railId, untilEpoch, false);
     }
@@ -1188,51 +1197,34 @@ contract Payments is
         Rail storage rail = rails[railId];
         RateChangeQueue.Queue storage rateQueue = rail.rateChangeQueue;
 
-        totalSettledAmount = 0;
-        totalNetPayeeAmount = 0;
-        totalPaymentFee = 0;
-        totalOperatorCommission = 0;
-        uint256 processedEpoch = startEpoch;
-        note = "";
+        SettlementState memory state = SettlementState({
+            totalSettledAmount: 0,
+            totalNetPayeeAmount: 0,
+            totalPaymentFee: 0,
+            totalOperatorCommission: 0,
+            processedEpoch: startEpoch,
+            note: ""
+        });
 
         // Process each segment until we reach the target epoch or hit an early exit condition
-        while (processedEpoch < targetEpoch) {
-            // Default boundary is the target we want to reach
-            uint256 segmentEndBoundary = targetEpoch;
-            uint256 segmentRate;
-
-            // If we have rate changes in the queue, use the rate from the next change
-            if (!rateQueue.isEmpty()) {
-                RateChangeQueue.RateChange memory nextRateChange = rateQueue
-                    .peek();
-
-                // Validate rate change queue consistency
-                require(
-                    nextRateChange.untilEpoch >= processedEpoch,
-                    "rate queue is in an invalid state"
-                );
-
-                // Boundary is the minimum of our target or the next rate change epoch
-                segmentEndBoundary = min(
-                    targetEpoch,
-                    nextRateChange.untilEpoch
-                );
-                segmentRate = nextRateChange.rate;
-            } else {
-                // If queue is empty, use the current rail rate
-                segmentRate = currentRate;
+        while (state.processedEpoch < targetEpoch) {
+            (uint256 segmentEndBoundary, uint256 segmentRate) = _getNextSegmentBoundary(
+                rateQueue,
+                currentRate,
+                state.processedEpoch,
+                targetEpoch
+            );
 
                 // if current rate is zero, there's nothing left to do and we've finished settlement
                 if (segmentRate == 0) {
                     rail.settledUpTo = targetEpoch;
                     return (
-                        totalSettledAmount,
-                        totalNetPayeeAmount,
-                        totalPaymentFee,
-                        totalOperatorCommission,
+                    state.totalSettledAmount,
+                    state.totalNetPayeeAmount,
+                    state.totalPaymentFee,
+                    state.totalOperatorCommission,
                         "Zero rate payment rail"
                     );
-                }
             }
 
             // Settle the current segment with potentially arbitrated outcomes
@@ -1244,43 +1236,43 @@ contract Payments is
                 string memory arbitrationNote
             ) = _settleSegment(
                     railId,
-                    processedEpoch,
+                    state.processedEpoch,
                     segmentEndBoundary,
                     segmentRate,
                     skipArbitration
                 );
 
             // If arbiter returned no progress, exit early without updating state
-            if (rail.settledUpTo <= processedEpoch) {
+            if (rail.settledUpTo <= state.processedEpoch) {
                 return (
-                    totalSettledAmount,
-                    totalNetPayeeAmount,
-                    totalPaymentFee,
-                    totalOperatorCommission,
+                    state.totalSettledAmount,
+                    state.totalNetPayeeAmount,
+                    state.totalPaymentFee,
+                    state.totalOperatorCommission,
                     arbitrationNote
                 );
             }
 
             // Add the settled amounts to our running totals
-            totalSettledAmount += segmentSettledAmount;
-            totalNetPayeeAmount += segmentNetPayeeAmount;
-            totalPaymentFee += segmentPaymentFee;
-            totalOperatorCommission += segmentOperatorCommission;
+            state.totalSettledAmount += segmentSettledAmount;
+            state.totalNetPayeeAmount += segmentNetPayeeAmount;
+            state.totalPaymentFee += segmentPaymentFee;
+            state.totalOperatorCommission += segmentOperatorCommission;
 
             // If arbiter partially settled the segment, exit early
             if (rail.settledUpTo < segmentEndBoundary) {
                 return (
-                    totalSettledAmount,
-                    totalNetPayeeAmount,
-                    totalPaymentFee,
-                    totalOperatorCommission,
+                    state.totalSettledAmount,
+                    state.totalNetPayeeAmount,
+                    state.totalPaymentFee,
+                    state.totalOperatorCommission,
                     arbitrationNote
                 );
             }
 
             // Successfully settled full segment, update tracking values
-            processedEpoch = rail.settledUpTo;
-            note = arbitrationNote;
+            state.processedEpoch = rail.settledUpTo;
+            state.note = arbitrationNote;
 
             // Remove the processed rate change from the queue
             if (!rateQueue.isEmpty()) {
@@ -1290,12 +1282,38 @@ contract Payments is
 
         // We've successfully settled up to the target epoch
         return (
-            totalSettledAmount,
-            totalNetPayeeAmount,
-            totalPaymentFee,
-            totalOperatorCommission,
-            note
+            state.totalSettledAmount,
+            state.totalNetPayeeAmount,
+            state.totalPaymentFee,
+            state.totalOperatorCommission,
+            state.note
         );
+    }
+
+    function _getNextSegmentBoundary(
+        RateChangeQueue.Queue storage rateQueue,
+        uint256 currentRate,
+        uint256 processedEpoch,
+        uint256 targetEpoch
+    ) internal view returns (uint256 segmentEndBoundary, uint256 segmentRate) {
+        // Default boundary is the target we want to reach
+        segmentEndBoundary = targetEpoch;
+        segmentRate = currentRate;
+
+        // If we have rate changes in the queue, use the rate from the next change
+        if (!rateQueue.isEmpty()) {
+            RateChangeQueue.RateChange memory nextRateChange = rateQueue.peek();
+
+            // Validate rate change queue consistency
+            require(
+                nextRateChange.untilEpoch >= processedEpoch,
+                "rate queue is in an invalid state"
+            );
+
+            // Boundary is the minimum of our target or the next rate change epoch
+            segmentEndBoundary = min(targetEpoch, nextRateChange.untilEpoch);
+            segmentRate = nextRateChange.rate;
+        }
     }
 
     function _settleSegment(

